src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
-                    int count = (int)o[1];
-                    byte[] dirsyncCookie = (byte[])o[2];
+                    dirsyncCookie = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out _);
 
-                    DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, (moreData == 0 ? false : true), count, controls[i].IsCritical, controls[i].GetValue());
+                    DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, moreData != 0, count, controls[i].IsCritical, value);
                     controls[i] = dirsync;
                 }
                 else if (controls[i].Type == "1.2.840.113556.1.4.474")
                 {
-                    // The control is a SortControl.
-                    int result;
+                    // The control is a SortControl, as described in RFC 2891.
+                    ResultCode result;
                     string attribute = null;
-                    object[] o = BerConverter.TryDecode("{ea}", value, out bool decodeSucceeded);
 
-                    // decode might fail as AD for example never returns attribute name, we don't want to unnecessarily throw and catch exception
-                    if (decodeSucceeded)
-                    {
-                        Debug.Assert(o != null && o.Length == 2);
-                        result = (int)o[0];
-                        attribute = (string)o[1];
-                    }
-                    else
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                    Debug.Assert(sequenceContentLength > 0);
+                    asnSpan = asnSpan.Slice(sequenceContentOffset);
+
+                    result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan, AsnEncodingRules.BER, out int bytesConsumed);
+                    asnSpan = asnSpan.Slice(bytesConsumed);
+
+                    // Attribute name is optional: AD for example never returns attribute name
+                    asnReadSuccessful = AsnDecoder.TryReadEncodedValue(asnSpan, AsnEncodingRules.BER, out Asn1Tag octetStringTag, out _, out int octetStringLength, out _);
+                    if (asnReadSuccessful)
                     {
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        Span<byte> attributeNameBuffer = octetStringLength <= AttributeNameStackAllocationThreshold ? attributeNameScratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
 
-                        result = (int)o[0];
+                        _ = AsnDecoder.TryReadOctetString(asnSpan, attributeNameBuffer, AsnEncodingRules.BER, out _, out _);
+                        attribute = s_utf8Encoding.GetString(attributeNameBuffer);
                     }
```

Comment:
TryReadEncodedValue can fail for a lot of sub-reasons; but I _think_ you're just using it for "there was still data".  In which case you should rewrite this to

```C#
if (!asnSpan.IsEmpty)
{
    scoped ReadOnlySpan<byte> attributeNameBuffer;

    if (asnSpan.Length <= AttributeNameStackAllocationThreshold)
    {
        asnReadSuccessful = AsnDecoder.TryReadOctetString(
            asnSpan,
            attributeNameScratchSpace,
            AsnEncodingRules.BER,
            out int consumed,
            out int written);

        // TryReadOctetString can't fail when the output buffer is at least as big as the input buffer.
        Debug.Assert(asnReadSuccessful);
        attributeNameBuffer = attributeNameScratchSpace.Slice(0, written);
        asnSpan = asnSpan.Slice(consumed);
    }
    else
    {
        attributeNameBuffer = AsnDecoder.ReadOctetString(
            asnSpan,
            AsnEncodingRules.BER,
            out int consumed);

        asnSpan = asnSpan.Slice(consumed);
    }

    attribute = s_utf8Encoding.GetString(attributeNameBuffer);
}

if (!asnSpan.IsEmpty)
{
    throw new BerConversionException();
}

SortResponseControl sort = ...;
```

Reply:
If this library wants to change to using array pool for "too big for the stack", then you'd go with something like your current code.  But once you're using `new byte[len]` you should just call the array-allocating variety.

Reply:
I'm using the pattern of `TryReadEncodedValue` + `TryReadOctetString` to do a few things. The octet strings in question are optional, so I'm interrogating the buffer to figure out it's about to yield an octet string and if so, how big it is. I'm also adding a post-hoc check to make sure that it's the last thing in the sequence, in line with the prior review comment.

Reply:
Let's assume you have a definite-length payload: `30 04 0A 01 01 04`.

* ReadSequence will report success.  The definite length encoding means it isn't going to bother checking that the children make sense.
* ReadEnumerated will consume 3 bytes (`0A 01 01`).  The sequence contents now have one byte left (which is never valid).
* TryReadEncodedValue will return false.  It can read a tag, but not a length, you're out of data.  Nothing was _corrupt_, so no exception, just a false return.
* What state are you in now?

The code I wrote is "there's more stuff.  Just read it as an OCTET STRING, because that's the only optional thing remaining".  That will cause an exception, because the data is insufficient and therefore corrupt.  If there was NO data remaining, it'd not read anything else, and satisfy the OPTIONAL.

I don't know what BerConverter would do in this invalid-data case.  If it previously rejected it (which I hope it would) then it needs to continue to be rejected.  If it's super loosey goosey, then what you have is more compatible (though I think the more strict read is still better)

Reply:
Your example SortControl payload and explanation makes sense. Incidentally, BerConverter will reject that if I use the `{eO}` format specifier.

That payload should naturally fail for either of two reasons:
* An octet string is expected to follow the enumerated value, and the data which actually follows that enumerated value isn't a valid octet string.
* The data following the enumerated value isn't an octet string, so we're expecting the enumerated value to be the last thing in the sequence. It isn't.

It doesn't right now because TryReadEncodedValue will return false and the updated code will ignore the extra data inside the payload's sequence. I need to fix that when handling PageControl, SortControl and VlvResponseControl.